mypy/checker.py

```diff
@@ -5595,6 +5595,55 @@ def has_no_custom_eq_checks(t: Type) -> bool:
                             ):
                                 if_map[operands[left_index]] = remove_optional(item_type)
+                        else:
+                            operand = operands[right_index]
+                            options: list[LiteralType] = []
+                            is_literal: bool = True
+
+                            # If we are checking against a Tuple[Literal[...]] *or* we are
+                            # checking against a literal list/set/tuple expression.
+                            if isinstance(iterable_type, TupleType):
+                                for ty in iterable_type.items:
+                                    if (
+                                        isinstance(ty, Instance)
+                                        and ty.last_known_value is not None
+                                    ):
+                                        options.append(ty.last_known_value)
+                                    else:
+                                        is_literal = False
+                                        break
+                            elif isinstance(operand, (ListExpr, SetExpr, TupleExpr)):
+                                # We are a literal expression!
+                                for item in operand.items:
+                                    ty = get_proper_type(self.lookup_type(item))
+                                    if (
+                                        isinstance(ty, Instance)
+                                        and ty.last_known_value is not None
+                                    ):
+                                        options.append(ty.last_known_value)
+                                    else:
+                                        is_literal = False
+                                        break
+
+                            if is_literal and len(options) > 1:
+                                if len(options) > 1:
+                                    ty = UnionType.make_union(options)
+                                else:
+                                    ty = options[0]
+
+                                yes, no = self.conditional_types_with_intersection(
+                                    item_type, [TypeRange(ty, is_upper_bound=False)], node
+                                )
+                                # Intersection caluclates UninhabitedType too
+                                # aggressively -- ignore them for now
+                                if yes is not None and not isinstance(
+                                    get_proper_type(yes), UninhabitedType
+                                ):
+                                    if_map[operands[left_index]] = yes
+                                if no is not None and not isinstance(
+                                    get_proper_type(no), UninhabitedType
+                                ):
+                                    else_map[operands[left_index]] = no
```

Comment:
I'm not sure if there's something else I should do here. I found that the `UninhabitedType` is often too aggressive (e.g. if `item_type=float` and `ty=Literal[1]`, then `yes = UninhabitedType`. While that's technically true (in that `float` and `Literal[1]` are incompatible types), at runtime that code can still run.